### PR TITLE
feat(ds4): ondisk prefix cache export/adopt; fix continued checkpoint keying

### PR DESCRIPTION
**Pull request by davide221**
## Summary

DeepSeek V4 snapshots were memory-only: the backend left `snapshot_ref` / `snapshot_adopt` at their defaults, so `DiskPrefixCache` detected a memory-only backend and disabled itself for this model (the gating added in #579). This PR implements the two hooks for DeepSeek V4 on the same shared disk cache used by Qwen3 / Qwen3.5 / Laguna, and fixes a keying inconsistency in the shared continued-checkpoint path.

### DeepSeek V4 ondisk prefix cache (`server/src/deepseek4`)
- Every snapshot tensor gets a stable name (`ds4_snap_*`), so the disk cache can serialize the context and rebind it after a reload.
- The host-side decode state (last logits, DSpark feature tail) and the per-layer compressed row counts travel as sidecar tensors in the same context. Variable lengths live in `ne[1]`, which the layout fingerprint normalizes, so snapshots of different lengths share one layout id.
- `deepseek4_snapshot_bind()` rebinds a deserialized context by name, with an optional per-shard prefix and shared-storage ownership.
- `DeepSeek4Backend::snapshot_ref` / `snapshot_adopt`: adopt validates against the compression schedule of the loaded weights (not the live cache), so a valid file is never dropped as corrupt while the target is parked. Paged concurrent serving stays memory-only, as in the Qwen3.5 backend.
- `DeepSeek4LayerSplitAdapter`: shard snapshots are merged into one named context (`ls<i>_` prefix plus adapter-level meta / HC state / prefill logits), mirroring the Qwen3.5 adapter. Mixed remote-shard splits stay memory-only.

### Continued checkpoint keying (`server/src/server/disk_prefix_cache.cpp`)
A disk entry is keyed by the hash of its prompt prefix and a hit restores the whole snapshot. `maybe_store_continued()` keyed files by the interval-aligned prefix while the snapshot extended to the generation's final position, so a later prompt sharing only the key prefix could restore a snapshot covering tokens it never contained. Now:
- continued checkpoints are keyed by the full token prefix the snapshot covers; the interval only paces them;
- `save()` refuses a snapshot whose position exceeds its key, and `read_file()` rejects such files so existing ones are dropped on lookup.

## Validation

Unit (lucebox5, HIP gfx1151;gfx1201 build on top of #598):
- `test_deepseek4_unit`: all cases pass, including new `test_monolithic_snapshot_disk_roundtrip` (real `DiskPrefixCache` save → lookup → adopt → restore → exact-hit decode, layout stability across lengths, rejection of a bogus context) and `test_layer_split_snapshot_disk_roundtrip`.
- `test_server_unit`: 453 passed, 0 failed, including new `test_disk_cache_rejects_snapshot_past_key` and `test_disk_cache_continued_keys_full_prefix`.

End to end (lucebox5, DeepSeek-V4-Flash-0731 ROCmFPX on R9700 + Strix, long-context profile, greedy, two-turn conversations of ~950 tokens, `--kv-cache-dir` set, prefix cache slots enabled):
- Files: 29.7 MB per 930-position snapshot; layout id identical across restarts.
- After a server restart, a new last turn on the cached conversation hits the disk entry (`--disk-prefix-cache auto`), is adopted, and prefills only the new turn: 17 to 28 tokens in 0.4 to 1.4 s (including the file read) versus 7 to 9 s for a cold prefill of ~950 tokens.
- Disk-restored outputs matched cold outputs in every cross-process comparison run (3/3, plus one repeat on the final build).

## Notes for reviewers
- `--disk-prefix-cache full` keys the exact full prompt, so a conversation with a new last turn misses; cross-session reuse needs `auto` (per request via `prefix_cache.scope`, or as the server default).
- DeepSeek chat markers produce a reusable boundary only after a completed assistant turn (BOS / EOS / User / Assistant markers), so single-turn prompts do not snapshot.
- Not covered: mixed remote-shard target split (memory-only), PFlash-compressed disk snapshots (DeepSeek V4 has no PFlash yet).

<a href="https://cubic.dev/pr/Luce-Org/lucebox/pull/716?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: server/src/deepseek4/deepseek4_layer_split_adapter.h [hunk 1]
```diff
@@ -107,8 +114,16 @@ class DeepSeek4LayerSplitAdapter : public LayerSplitAdapter {
         std::vector<float> prefill_last_logits;
         std::vector<DeepSeek4Snapshot> shards;
         bool used = false;
+        // Merged serialization context (owned here). After snapshot_save it
+        // is a copy of the shard snapshots; after snapshot_adopt the shard
+        // snapshots alias it (owns_storage=false).
+        ggml_context *        disk_ctx = nullptr;
+        ggml_backend_buffer_t disk_buf = nullptr;
+        ggml_backend_t        disk_backend = nullptr;  // CPU backend owning disk_buf (save path)
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/src/deepseek4/deepseek4_layer_split_adapter.h]: Addressed in d0979d44: the reader now allocates from the CPU buffer type (no backend object to hand over), and the adapter keeps one merged context per slot that it owns and frees in snapshot_free().

File: server/src/deepseek4/deepseek4_internal.h [hunk 2]
```diff
@@ -411,11 +413,29 @@ bool build_deepseek4_head4_tail2_routes(
...
+                             const DeepSeek4SnapshotAux * aux = nullptr);
 bool deepseek4_snapshot_restore(const DeepSeek4Snapshot & snap,
                                 DeepSeek4Cache & cache);
+// Rebind a deserialized snapshot context (tensors named as written by
+// deepseek4_snapshot_save, optionally prefixed by `name_prefix`) into `out`.
+// Requires the meta tensor; fills row counts and cur_pos from it and
+// validates that every tensor referenced by the meta exists with a sane
+// shape. `out` takes ownership of ctx/buf iff `take_ownership`. Returns
+// false (and leaves `out` empty) on any inconsistency.
+bool deepseek4_snapshot_bind(ggml_context * ctx,
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/src/deepseek4/deepseek4_internal.h]: Addressed in d0979d44: adopt runs deepseek4_snapshot_validate(), which checks every tensor's type and shape against the geometry implied by the weights (plus capacity when the live cache exists). A mismatching file is rejected on lookup and removed; both paths now have tests with a corrupted tensor.

File: server/src/deepseek4/deepseek4_layer_split_adapter.cpp [hunk 3]
```diff
@@ -658,6 +664,224 @@ bool DeepSeek4LayerSplitAdapter::snapshot_save(int slot) {
...
+    ggml_tensor * logits = ggml_new_tensor_2d(
+        ctx, GGML_TYPE_F32, 1, (int64_t) std::max<size_t>(1, snap.prefill_last_logits.size()));
+    if (!meta || !hc || !logits) { ggml_free(ctx); return false; }
+    ggml_set_name(meta, kDs4SplitMetaName);
+    ggml_set_name(hc, kDs4SplitHcName);
+    ggml_set_name(logits, kDs4SplitLogitsName);
+
+    ggml_backend_t cpu = ggml_backend_cpu_init();
+    if (!cpu) { ggml_free(ctx); return false; }
+    ggml_backend_buffer_t buf = ggml_backend_alloc_ctx_tensors(ctx, cpu);
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/src/deepseek4/deepseek4_layer_split_adapter.cpp]: Addressed in d0979d44: the adapter no longer keeps two copies. Shard snapshots are declared into a single merged host context (aliased, owns_storage=false) that serves both in-memory restore and disk export; rebuild_disk_snapshot is gone.

File: server/test/test_server_unit.cpp [hunk 4]
```diff
@@ -5030,10 +5030,115 @@ static std::array<uint8_t, 16> read_layout_id_from_cache_dir(const std::string &
...
+    for (int i = 0; i < 40; ++i) tokens.push_back(100 + i);
+    const int cur_pos = MockBackendWithLayout::kMaxPos;  // 32
+    TEST_ASSERT(cache.maybe_store_continued(0, tokens, cur_pos));
+    std::vector<int32_t> aligned(tokens.begin(), tokens.begin() + 30);
+    std::vector<int32_t> covered(tokens.begin(), tokens.begin() + cur_pos);
+    TEST_ASSERT(!cache.lookup(aligned, 1));
+    TEST_ASSERT(cache.lookup(covered, 1));
+    TEST_ASSERT(backend.adopted_cur_pos_ == cur_pos);
+    // Same interval bucket: no second checkpoint.
+    TEST_ASSERT(!cache.maybe_store_continued(0, tokens, cur_pos));
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/test/test_server_unit.cpp:5146]: Addressed in d0979d44: the test now advances the mock position across a new interval bucket (32 -> 42), expects a second checkpoint keyed by the 42 covered tokens, checks that the aligned 40-token key misses while the earlier checkpoint still hits, and that a later position in the same bucket does not refire.

File: server/tests/test_deepseek4_unit.cpp [hunk 5]
```diff
@@ -2176,6 +2177,343 @@ static void test_monolithic_snapshot_preserves_decode_state() {
...
+    return true;
+}
+
+static std::string make_test_disk_cache_dir(const char * tag) {
+    return "/tmp/dflash_test_ds4_disk_" + std::string(tag) + "_" +
+           std::to_string((long) getpid());
+}
+
+static void remove_test_disk_cache_dir(const std::string & dir) {
+    const std::string cmd = "rm -rf '" + dir + "'";
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/tests/test_deepseek4_unit.cpp]: Addressed in d0979d44: std::filesystem::remove_all replaces the shell call.

File: server/src/deepseek4/deepseek4_layer_split_adapter.cpp [hunk 6]
```diff
@@ -636,23 +639,113 @@ bool DeepSeek4LayerSplitAdapter::decode_ar(
...
+        snap.shards[i].ctx = ctx;
+        snap.shards[i].buf = buf;
+        if (!deepseek4_snapshot_fill(shards_[i].cache, &empty_aux, snap.shards[i])) {
+            for (auto & shard_snap : snap.shards) shard_snap = DeepSeek4Snapshot{};
+            ggml_backend_buffer_free(buf);
+            ggml_free(ctx);
             return false;
         }
     }
     if (use_mixed_target_split() && !remote_target_shard_.snapshot_save(slot)) {
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/src/deepseek4/deepseek4_layer_split_adapter.cpp:734]: Addressed in 21d4be7a: the slot takes ownership of the merged context right after allocation, and every later failure path (shard fill, remote shard save) goes through snapshot_free(), which releases the local context and the remote shard's slot together.

File: server/src/deepseek4/deepseek4_snapshot.cpp [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+                             DeepSeek4SnapshotBindInfo * info) {
+    free_deepseek4_snapshot(out);
+    if (!ctx || !buf) return false;
+
+    ggml_tensor * meta = find_named(ctx, deepseek4_snapshot_tensor_name(name_prefix, kDeepSeek4SnapMetaName));
+    if (!meta || meta->type != GGML_TYPE_I32 || ggml_n_dims(meta) != 1 ||
+        meta->ne[0] < kDeepSeek4SnapMetaBase || !meta->data) {
+        return false;
+    }
+    std::vector<int32_t> m((size_t) meta->ne[0], 0);
```
[the remark this answers is not in the document]
Comment (davide221) [reply server/src/deepseek4/deepseek4_snapshot.cpp:494]: Addressed in 21d4be7a: the meta length is bounded by the 4096-layer limit before the vector is allocated.
